File: android/app/src/main/java/com/retroportalstudio/www/background_service/NotificationBase.java
package com.retroportalstudio.www.background_service;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

public class NotificationBase {
}

class WPNotification {
    String url =
            "https://messaging.dinavision.org/api/v1/Messaging/SendByUserName";
    String username;
    NotificationBody data;

    WPNotification(){

    }

    WPNotification(String un, NotificationBody nb){
        username = un;
        data = nb;
    }

    WPNotification fromJson(JSONObject jsn) {
        try{
            this.url = jsn.getString("url");
            this.username = jsn.getString("username");
            this.data = new NotificationBody().fromJson(jsn.getJSONObject("data"));

            return this;
        }catch (Exception Ex){
            return this;
        }
    }
}

class NotificationBody {
    WebPushNotification notification;

    NotificationBody(){

    }

    NotificationBody(WebPushNotification not){
        notification = not;
    }

    NotificationBody fromJson(JSONObject jsn) {
        try{
            this.notification = new WebPushNotification().fromJson(jsn.getJSONObject("notification"));

            return this;
        }catch (Exception Ex){
            Log.d("error", Ex.getMessage());
        }

        return this;
    }
}

class WebPushNotification {
    String title;
    String body;
    String tag;
    String icon;
    String image;
    String badge;
    String vibrate;
    String sound;
    String dir;
    Boolean renotify = true;
    Boolean requireInteraction;
    Boolean silent;
    ArrayList<WebNotificationAction> actions;
    int timestamp;
    WebPushNotificationData data;

    WebPushNotification fromJson(JSONObject jsn){
        try{
            this.title = jsn.getString("title");
            this.body = jsn.getString("body");
            this.tag = jsn.getString("tag");
            this.icon = jsn.getString("icon");
            this.image = jsn.getString("image");
            this.badge = jsn.getString("badge");
            this.vibrate = jsn.getString("vibrate");
            this.sound = jsn.getString("sound");
            this.dir = jsn.getString("dir");
            this.renotify = jsn.getBoolean("renotify");
            this.requireInteraction = jsn.getBoolean("requireInteraction");
            this.silent = jsn.getBoolean("silent");
            this.actions = new WebNotificationAction().listFromJson(jsn.getJSONArray("actions"));
            this.timestamp = Integer.parseInt(jsn.getString("timestamp"));
            this.data = new WebPushNotificationData().fromJson(jsn.getJSONObject("data"));

            return this;
        }catch (Exception Ex){
            Log.d("error", Ex.getMessage());
        }

        return this;
    }
}

class WebPushNotificationData {
    String time;
    String message;

    WebPushNotificationData(){

    }

    WebPushNotificationData(String tm, String msg){
        message = msg;
        time = tm;
    }

    WebPushNotificationData fromJson(JSONObject jsn){
        try{
            this.message = jsn.getString("message");
            this.time = jsn.getString("time");
        }catch (Exception Ex){
            Log.d("error", Ex.getMessage());
        }
        
        return this;
    }
}

class WebNotificationAction {
    String action;
    String title;
    String icon;

    WebNotificationAction(){

    }

    WebNotificationAction(String ic, String ttl, String act){
        action = act;
        title = ttl;
        icon = ic;
    }

    WebNotificationAction fromJson(JSONObject jsn){
        try{
            this.icon = jsn.getString("icon");
            this.title = jsn.getString("title");
            this.action = jsn.getString("action");
            
            return this;
        }catch (Exception Ex){
            Log.d("error", Ex.getMessage());
        }

        return this;
    }

    ArrayList<WebNotificationAction> listFromJson(JSONArray jsn){
        try{
            ArrayList<WebNotificationAction> list = new ArrayList<WebNotificationAction>();
            if(jsn != null)
            {
                for (int i=0;i<jsn.length();i++){
                    list.add(new WebNotificationAction().fromJson(jsn.getJSONObject(i)));
                }
            }

            return list;
        }catch (Exception Ex){
            Log.d("error", Ex.getMessage());
        }

        return null;
    }
}

class NotificationResult {
    Boolean isSuccess;
    String statusCode;
    String message;
}

File: lib/NotificationBase.dart
class WPNotification {
  String url =
      "https://messaging.dinavision.org/api/v1/Messaging/SendByUserName";
  String username;
  NotificationBody data;

  WPNotification({this.data, this.username});

  WPNotification fromJson(dynamic jsn) {
    this.url = jsn["url"].toString();
    this.username = jsn["username"].toString();
    this.data = NotificationBody().fromJson(jsn["data"]);

    return this;
  }
}

class NotificationBody {
  WebPushNotification notification;

  NotificationBody({this.notification});

  NotificationBody fromJson(dynamic jsn) {
    this.notification = jsn["notification"];

    return this;
  }
}

class WebPushNotification {
  String title;
  String body;
  String tag;
  String icon;
  String image;
  String badge;
  String vibrate;
  String sound;
  String dir;
  bool renotify = true;
  bool requireInteraction;
  bool silent;
  List<WebNotificationAction> actions;
  int timestamp;
  WebPushNotificationData data;

  WebPushNotification fromJson(dynamic jsn){
    this.title = jsn["title"].toString();
    this.body = jsn["body"].toString();
    this.tag = jsn["tag"].toString();
    this.icon = jsn["icon"].toString();
    this.image = jsn["image"].toString();
    this.badge = jsn["badge"].toString();
    this.vibrate = jsn["vibrate"].toString();
    this.sound = jsn["sound"].toString();
    this.dir = jsn["dir"].toString();
    this.renotify = jsn["renotify"];
    this.requireInteraction = jsn["requireInteraction"];
    this.silent = jsn["silent"];
    this.actions = WebNotificationAction().listFromJson(jsn["actions"]);
    this.timestamp = int.parse(jsn["timestamp"].toString());
    this.data = WebPushNotificationData().fromJson(jsn["data"]);

    return this;
  }
}

class WebPushNotificationData {
  String time;
  String message;

  WebPushNotificationData({
    this.time,
    this.message
});

  WebPushNotificationData fromJson(dynamic jsn){
    this.message = jsn["message"].toString();
    this.time = jsn["time"].toString();

    return this;
  }
}

class WebNotificationAction {
  String action;
  String title;
  String icon;

  WebNotificationAction({this.title, this.action, this.icon});

  WebNotificationAction fromJson(dynamic jsn){
    this.icon = jsn["icon"].toString();
    this.title = jsn["title"].toString();
    this.action = jsn["action"].toString();

    return this;
  }

  List<WebNotificationAction> listFromJson(dynamic jsn){
    if(jsn != null)
    {
      return jsn.map((data) => WebNotificationAction().fromJson(data))
          .toList();
    }

    return null;
  }
}

class NotificationResult {
  bool isSuccess;
  String statusCode;
  String message;
}
